File: examples/datasets/groundstations_coverage.py
# /// script
# dependencies = ["brahe"]
# ///
"""
Network coverage analysis for groundstation datasets.

Demonstrates:
- Loading all providers
- Analyzing by latitude band
- Finding stations by capability
"""

import brahe as bh

if __name__ == "__main__":
    print("Groundstation Network Coverage Analysis")
    print("=" * 60)

    # Load all providers
    all_stations = bh.datasets.groundstations.load_all()
    print(f"\nTotal stations across all providers: {len(all_stations)}")

    # Analyze by latitude band
    arctic = [s for s in all_stations if s.lat > 66.5]
    temperate = [s for s in all_stations if -66.5 <= s.lat <= 66.5]
    antarctic = [s for s in all_stations if s.lat < -66.5]

    print("\nLatitude distribution:")
    print(f"  Arctic stations (>66.5°N): {len(arctic)}")
    print(f"  Temperate stations: {len(temperate)}")
    print(f"  Antarctic stations (<66.5°S): {len(antarctic)}")

    # Find stations by capability
    x_band_stations = [
        s for s in all_stations if "X" in s.properties["frequency_bands"]
    ]
    print("\nCapability analysis:")
    print(f"  X-band capable stations: {len(x_band_stations)}")

    print("\n" + "=" * 60)
    print("Coverage analysis complete!")

File: examples/datasets/groundstations_coverage.rs
//! Network coverage analysis for groundstation datasets.
//!
//! Demonstrates:
//! - Loading all providers
//! - Analyzing by latitude band
//! - Finding stations by capability

use brahe::datasets::groundstations;
use brahe::access::AccessibleLocation;

fn main() {
    println!("Groundstation Network Coverage Analysis");
    println!("{}", "=".repeat(60));

    // Load all providers
    let all_stations = groundstations::load_all_groundstations().unwrap();
    println!("\nTotal stations across all providers: {}", all_stations.len());

    // Analyze by latitude band
    let arctic: Vec<_> = all_stations.iter().filter(|s| s.lat() > 66.5).collect();
    let temperate: Vec<_> = all_stations.iter().filter(|s| s.lat() >= -66.5 && s.lat() <= 66.5).collect();
    let antarctic: Vec<_> = all_stations.iter().filter(|s| s.lat() < -66.5).collect();

    println!("\nLatitude distribution:");
    println!("  Arctic stations (>66.5°N): {}", arctic.len());
    println!("  Temperate stations: {}", temperate.len());
    println!("  Antarctic stations (<66.5°S): {}", antarctic.len());

    // Find stations by capability (check if "X" appears in frequency_bands array)
    let x_band_stations: Vec<_> = all_stations.iter()
        .filter(|s| {
            if let Some(bands) = s.properties().get("frequency_bands") {
                bands.to_string().contains("X")
            } else {
                false
            }
        })
        .collect();

    println!("\nCapability analysis:");
    println!("  X-band capable stations: {}", x_band_stations.len());

    println!("\n{}", "=".repeat(60));
    println!("Coverage analysis complete!");
}
